source/opt/interface_repair_pass.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "interface_repair_pass.h"
+#include "source/spirv_constant.h"
+namespace spvtools {
+namespace opt {
+InterfaceRepairPass::Status InterfaceRepairPass::Process() {
+  std::unordered_map<uint32_t, Function*> id2function_;
+  for (auto& fn : *get_module()) id2function_[fn.result_id()] = &fn;
+  bool modified = false;
+  for (auto& entry : get_module()->entry_points()) {
+    std::set<uint32_t> used_variables;
+    std::function<void(Function * func)> traverser = [&](Function* func) {
+      for (auto bi = func->begin(); bi != func->end(); ++bi) {
+        for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+          if (ii->opcode() == SpvOpFunctionCall) {
+            const auto called_id = ii->GetOperand(2).words[0];
+            traverser(id2function_.at(called_id));
```

Comment:
You don't have to traverse the call graph yourself.  We have utility functions that can do that for you.  See `IRContext::ProcessCallTreeFromRoots` and its uses for examples.